### DISCUSSION TITLE
MarkovJunior Interactive Extensions

### DISCUSSION THREAD

**Discussion by Yuu6883** [mxgmn/MarkovJunior#46] [Show and tell]
I modified the original Tetris model and extended a few nodes and recreated the Tetris game with MarkovJunior ([live demo](https://yuu6883.github.io/MarkovJuniorWeb/?model=Tetris))
![tetris](https://user-images.githubusercontent.com/38842891/179383760-f4409095-de0d-4d58-8876-797619c8f196.gif)

[Model file](https://github.com/Yuu6883/MarkovJuniorWeb/blob/main/static/models/Tetris.xml)
Extended nodes for better control flow of the program:
- `<scope>` node is like `<sequence>`, but more like a passthrough like a { scope block } in traditional language
- `<and>` node runs its children sequentially, but terminates when a child node fails to match anything, useful for a condition chain in the program.
- `<event>` node always return success but fail after a certain amount of time pass, it also runs its <event-handler> children which would run their subtree if the event condition matches, e.g. key down/up.

There is a problem with symmetry that's not matching actual Tetris but can be fixed by splitting the rules (left L shape and right L shape with 2 different colors).

**Comment by mxgmn** [comment]
Impressive! I see it came a long way. The speed slider doesn't really increase speed x100 for me, although in other (non-interactive?) models it does.

**Reply by Yuu6883** [reply]
Oh yeah I forgot to scale the timeout value in the event node with the speed.

**Comment by alan2here** [comment]
The sequential Sokoban is the only Sokoban one that does the right thing and doesn't lock up the webpage so it needs to be reloaded.

**Reply by Yuu6883** [reply]
hmm what browser are you using? All the Sokoban models work fine for me on chrome.

**Reply by alan2here** [reply]
IOS Chrome, which uses the components IOS Safari uses.

**Reply by Yuu6883** [reply]
![00ABDCE3-6F63-4897-8ADB-9ADD2D9E137E](https://user-images.githubusercontent.com/38842891/182974262-395ee5db-a639-440e-a02c-8430e26f3605.png)
Works fine on my iPhone, maybe the site is cached to some older version that tries to solve it synchronously?

**Reply by alan2here** [reply]
https://user-images.githubusercontent.com/4549363/182975245-f8893ef5-5070-4ce0-9bf9-c419b8fdb425.MOV

This at least tries to solve it, but still leaps 10s of moves each time and then gets stuck, the others just scatter the boxes around. It's on IOS Safari but they both work as well as each other.

**Reply by Yuu6883** [reply]
Well changing the speed to 100ms would make it slower, leaving it at 1x is fine and there's no need to pause it. The canvas only updates when a closer-to-solution state is found so it might seem stuck when searching the states. SokobanLevel1 usually searches around 300-400k states which might take at least 30s on my iPhone 10 and possibly longer on older devices. SokobanLevel2 should run really fast (less than a second). I wouldn't recommend running the multi ones on mobile lol. One of them might search from 500k to a few million states and takes maybe 1GB memory and it would take forever to run on mobile.

**Reply by alan2here** [reply]
For comparison:

https://user-images.githubusercontent.com/4549363/183040197-4965ead5-ae99-482a-a94c-38277b0cc6c7.MOV

Clearly runs more smoothly, doesn't skip frames, every move shown.

---

Also:

https://user-images.githubusercontent.com/4549363/183043852-e7f174b7-4819-4231-80c2-440b8a27e6f0.MOV

No problems with performance on this device, in general which is more than can be said for a lot of PCs, although The Witness seems well optimised enough to generally work.

**Reply by Yuu6883** [reply]
SequentialSokoban only searches about 400 states while SokobanLevel1 searches 400k states... (so it's supposed to run 1000x slower). This algorithm also only run in single thread on CPU, so you can't compare a completely different thing like a 3D renderer which probably runs on multiple threads and mostly uses GPU to render.

**Reply by alan2here** [reply]
I guess then level 1 is supposed to have a few single steps where many steps of movement happen all at once. That's why I slowed it down with the 100ms delay, I was planning to watch the path that it pushed the box along.

It's nice to know that it eventually gets there with that final box.

**Reply by Yuu6883** [reply]
The actual path will be shown after the search completes
